src/jit/codegenxarch.cpp

```diff
@@ -6222,10 +6222,27 @@ void CodeGen::genCompareInt(GenTreePtr treeNode)
     assert(!varTypeIsFloating(op2Type));
 
     instruction ins;
+    var_types   type = TYP_UNKNOWN;
 
     if (tree->OperIs(GT_TEST_EQ, GT_TEST_NE))
     {
         ins = INS_test;
+
+        // Unlike many xarch instructions TEST doesn't have a form with a 16/32/64 bit first operand and
+        // an 8 bit immediate second operand. But if the immediate value fits in 8 bits then we can simply
+        // emit a 8 bit TEST instruction, unless we're targeting x86 and the first operand is a non-byteable
+        // register.
+        // Note that lowering does something similar but its main purpose is to allow memory operands to be
+        // contained so it doesn't handle other kind of operands. It could do more but on x86 that results
+        // in additional register constrains and that may be worse than wasting 3 bytes on an immediate.
+        if (
+#ifdef _TARGET_X86_
+            (!op1->isUsedFromReg() || isByteReg(op1->gtRegNum)) &&
+#endif
+            (op2->IsCnsIntOrI() && genTypeCanRepresentValue(TYP_UBYTE, op2->AsIntCon()->IconValue())))
+        {
+            type = TYP_UBYTE;
+        }
```

Comment:
> Do we ever issue test r/m64, imm32 today?

Yes.

> That particular encoding sign-extends the immediate and thus could not be replaced with test r/m8, imm8.

I don't quite follow. This transform is done only if the immediate value is in range 0..255. So if any sign extension is/would be done anywhere it doesn't matter since the value is positive.